src/go/rpk/pkg/config/params.go

```diff
@@ -1612,26 +1612,101 @@ func (c *Config) addUnsetRedpandaDefaults(actual bool) {
 		src.Redpanda.AdminAPITLS,
 		&dst.Rpk.AdminAPI.Addresses,
 	)
-
-	if len(dst.Rpk.KafkaAPI.Brokers) == 0 && len(dst.Rpk.AdminAPI.Addresses) > 0 {
-		_, host, _, err := rpknet.SplitSchemeHostPort(dst.Rpk.AdminAPI.Addresses[0])
-		if err == nil {
-			host = net.JoinHostPort(host, strconv.Itoa(DefaultKafkaPort))
-			dst.Rpk.KafkaAPI.Brokers = []string{host}
-			dst.Rpk.KafkaAPI.TLS = dst.Rpk.AdminAPI.TLS
+	if src.SchemaRegistry != nil {
+		defaultFromRedpanda(
+			namedAuthnToNamed(src.SchemaRegistry.SchemaRegistryAPI),
+			src.SchemaRegistry.SchemaRegistryAPITLS,
+			&dst.Rpk.SR.Addresses,
+		)
+	}
+
+	// If only one listener is set, infer the others using its address, TLS
+	// config, and default ports. This isn't perfect, but it's a reasonable
+	// fallback.
+
+	// This list defines a set of sources we can use to infer the others.
+	// Each entry includes the source address list, its TLS config, and a list
+	// of "setters" — where each setter is a target to populate if it's empty.
+	srcs := []struct {
+		from    []string
+		tls     *TLS
+		setters []struct {
```

Comment:
preference nit: I'd probably unnest this struct into three fields since you have to define the shape in every iteration below

Reply:
Could do something like,

`srcAddrs, srcTLS, dstAddrs, dstTLS, defPort`

Reply:
Ended up doing something in between. Defined the type above so I don't have to repeat it every time. And also changed the names.

I didn't flatten the struct though